commit 73208971e5e9a11d4de14363a3f50bf2b45ab727
Author: Liping Zhang <zlpnobody@gmail.com>
Date:   Fri Apr 7 23:51:07 2017 +0800

    sysctl: report EINVAL if value is larger than UINT_MAX for proc_douintvec
    
    commit 425fffd886bae3d127a08fa6a17f2e31e24ed7ff upstream.
    
    Currently, inputting the following command will succeed but actually the
    value will be truncated:
    
      # echo 0x12ffffffff > /proc/sys/net/ipv4/tcp_notsent_lowat
    
    This is not friendly to the user, so instead, we should report error
    when the value is larger than UINT_MAX.
    
    Fixes: e7d316a02f68 ("sysctl: handle error writing UINT_MAX to u32 fields")
    Signed-off-by: Liping Zhang <zlpnobody@gmail.com>
    Cc: Subash Abhinov Kasiviswanathan <subashab@codeaurora.org>
    Cc: Andrew Morton <akpm@linux-foundation.org>
    Cc: Eric W. Biederman <ebiederm@xmission.com>
    Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

diff --git a/kernel/sysctl.c b/kernel/sysctl.c
index 7e97947..bc15d6c 100644
--- a/kernel/sysctl.c
+++ b/kernel/sysctl.c
@@ -2158,6 +2158,8 @@ static int do_proc_douintvec_conv(bool *negp, unsigned long *lvalp,
 	if (write) {
 		if (*negp)
 			return -EINVAL;
+		if (*lvalp > UINT_MAX)
+			return -EINVAL;
 		*valp = *lvalp;
 	} else {
 		unsigned int val = *valp;
